File: freud/py/triage.py
#!/usr/bin/env python3
"""Heuristic triage of a Freud run directory.

Port of freud/scripts/triage-run.sh — scans logs for known anomaly patterns,
pressure spikes, and top error/warning signals. Produces:
  artifacts/anomalies.json
  artifacts/anomalies.md
  artifacts/pattern-counts.tsv
  artifacts/top-signals.tsv
  artifacts/pressure-signals.tsv
  artifacts/first-failing-trace.txt
"""

from __future__ import annotations

import argparse
import os
import re
import sys
from pathlib import Path

from freud.py.common import (
    json_escape,
    non_empty_lines,
    rg_search,
    utcnow_iso,
)

PATTERNS: list[tuple[str, str]] = [
    ("planner_output_repaired", "planner_output_repaired|output_repaired"),
    ("parse_failures", "non-parseable|failed to parse|parse fallback|parse error|parse failure"),
    ("forced_terminal", "forced terminal|forced_terminal_answer"),
    ("queue_saturation", "queue full|queue_saturation|step limit reached"),
    ("policy_denials", "action denied|denied by superego|policy denied"),
    ("provider_failures", "timeout|unavailable|provider check failed"),
    ("cache_divergence", "llm_cache_divergence"),
]



def _record_pattern(pattern_id: str, regex: str, logs_dir: Path) -> tuple[str, int, str]:
    """Search logs for *regex* and return (pattern_id, count, first_sample)."""
    if not logs_dir.is_dir():
        return (pattern_id, 0, "")
    raw = rg_search(regex, logs_dir, case_insensitive=True)
    lines = non_empty_lines(raw)
    count = len(lines)
    sample = lines[0] if lines else ""
    return (pattern_id, count, sample)


def _extract_pressure_value(line: str) -> str | None:
    m = re.search(r"decision_pressure=([0-9]+\.[0-9]+)", line)
    return m.group(1) if m else None


def _compare_gt(a: str, b: str) -> bool:
    try:
        return float(a) > float(b)
    except ValueError:
        return False


def _compare_ge(a: str, threshold: float) -> bool:
    try:
        return float(a) >= threshold
    except ValueError:
        return False


def _context_around(filepath: str, line_num: int, context: int = 2) -> str:
    """Read a few lines of context around *line_num* from *filepath*."""
    try:
        all_lines = Path(filepath).read_text(encoding="utf-8", errors="replace").splitlines()
    except OSError:
        return ""
    start = max(0, line_num - 1 - context)
    end = min(len(all_lines), line_num + context)
    numbered = []
    for i, l in enumerate(all_lines[start:end], start=start + 1):
        numbered.append(f"  {i}\t{l}")
    return "\n".join(numbered)


def triage(run_dir: str, *, top_n: int = 20) -> str:
    """Run triage on *run_dir*. Returns path to anomalies.json."""
    run_path = Path(run_dir)
    if not run_path.is_dir():
        print(f"Run directory does not exist: {run_dir}", file=sys.stderr)
        sys.exit(1)

    logs_dir = run_path / "logs"
    artifact_dir = run_path / "artifacts"
    artifact_dir.mkdir(parents=True, exist_ok=True)

    patterns_file = artifact_dir / "pattern-counts.tsv"
    top_signals_file = artifact_dir / "top-signals.tsv"
    pressure_file = artifact_dir / "pressure-signals.tsv"
    first_failure_file = artifact_dir / "first-failing-trace.txt"

    # --- record patterns ---
    pattern_results: list[tuple[str, int, str]] = []
    for pid, regex in PATTERNS:
        pattern_results.append(_record_pattern(pid, regex, logs_dir))

    with open(patterns_file, "w", encoding="utf-8") as f:
        for pid, count, sample in pattern_results:
            f.write(f"{pid}\t{count}\t{sample}\n")

    # --- top signals & pressure ---
    top_lines: list[str] = []
    pressure_lines: list[str] = []
    first_fail_line = ""

    if logs_dir.is_dir():
        raw_top = rg_search("error|exception|failed|warning", logs_dir, case_insensitive=True,
                            invert_match=":> Task :")
        top_lines = non_empty_lines(raw_top)[:top_n]

        raw_pressure = rg_search(r"decision_pressure=[0-9]+\.[0-9]+", logs_dir,
                                 case_insensitive=False)
        pressure_lines = non_empty_lines(raw_pressure)

        raw_fail = rg_search("error|exception|failed|traceback|assert", logs_dir,
                             case_insensitive=True, invert_match=":> Task :")
        fail_lines = non_empty_lines(raw_fail)
        first_fail_line = fail_lines[0] if fail_lines else ""

    with open(top_signals_file, "w", encoding="utf-8") as f:
        f.write("\n".join(top_lines) + ("\n" if top_lines else ""))

    with open(pressure_file, "w", encoding="utf-8") as f:
        f.write("\n".join(pressure_lines) + ("\n" if pressure_lines else ""))

    # --- first failing trace snippet ---
    first_failure_text = ""
    if first_fail_line:
        parts = first_fail_line.split(":", 2)
        file_ref = parts[0] if len(parts) > 0 else ""
        line_rest = parts[1] if len(parts) > 1 else ""
        if os.path.isfile(file_ref) and line_rest.isdigit():
            line_num = int(line_rest)
            ctx = _context_around(file_ref, line_num)
            first_failure_text = f"{first_fail_line}\n---\n{ctx}"
        else:
            first_failure_text = first_fail_line

    with open(first_failure_file, "w", encoding="utf-8") as f:
        f.write(first_failure_text)

    # --- pressure analysis ---
    max_pressure_val = ""
    max_pressure_ref = ""
    first_pressure_075 = ""
    first_pressure_090 = ""

    for line in pressure_lines:
        value = _extract_pressure_value(line)
        if value is None:
            continue
        if not max_pressure_val or _compare_gt(value, max_pressure_val):
            max_pressure_val = value
            max_pressure_ref = line
        if not first_pressure_075 and _compare_ge(value, 0.75):
            first_pressure_075 = line
        if not first_pressure_090 and _compare_ge(value, 0.90):
            first_pressure_090 = line

    # --- anomalies.json ---
    anomalies_json_path = artifact_dir / "anomalies.json"
    generated_at = utcnow_iso()

    pattern_entries = []
    for pid, count, sample in pattern_results:
        pattern_entries.append(
            f'    {{"id":"{json_escape(pid)}","count":{count},"sample":"{json_escape(sample)}"}}'
        )

    top_signal_entries = []
    for line in top_lines:
        if line:
            top_signal_entries.append(f'    "{json_escape(line)}"')

    json_lines = [
        "{",
        f'  "generated_at": "{generated_at}",',
        f'  "run_dir": "{json_escape(run_dir)}",',
        '  "pattern_counts": [',
        ",\n".join(pattern_entries),
        "  ],",
        '  "pressure": {',
        f'    "max": {{"value": "{json_escape(max_pressure_val)}", "ref": "{json_escape(max_pressure_ref)}"}},',
        f'    "first_ge_075": "{json_escape(first_pressure_075)}",',
        f'    "first_ge_090": "{json_escape(first_pressure_090)}"',
        "  },",
        f'  "first_failing_trace": "{json_escape(first_failure_text)}",',
        '  "top_signals": [',
        ",\n".join(top_signal_entries),
        "  ]",
        "}",
    ]
    anomalies_json_path.write_text("\n".join(json_lines) + "\n", encoding="utf-8")

    # --- anomalies.md ---
    anomalies_md_path = artifact_dir / "anomalies.md"
    md_lines = [
        "# Anomaly Triage",
        "",
        f"- Run dir: `{run_dir}`",
        f"- Generated at: `{generated_at}`",
        "",
        "## Pattern Counts",
    ]
    for pid, count, sample in pattern_results:
        md_lines.append(f"- `{pid}`: {count}")
        if sample:
            md_lines.append(f"  sample: `{sample}`")
    md_lines.append("")
    md_lines.append("## Pressure Spikes")
    md_lines.append(f"- max decision pressure: `{max_pressure_val or 'none'}`")
    if max_pressure_ref:
        md_lines.append(f"  ref: `{max_pressure_ref}`")
    if first_pressure_075:
        md_lines.append(f"- first >= 0.75: `{first_pressure_075}`")
    else:
        md_lines.append("- first >= 0.75: none")
    if first_pressure_090:
        md_lines.append(f"- first >= 0.90: `{first_pressure_090}`")
    else:
        md_lines.append("- first >= 0.90: none")
    md_lines.append("")
    md_lines.append("## First Failing Trace Snippet")
    if first_failure_text:
        md_lines.append("```text")
        md_lines.append(first_failure_text)
        md_lines.append("```")
    else:
        md_lines.append("- none")
    md_lines.append("")
    md_lines.append("## Top Signals")
    if top_lines:
        for line in top_lines:
            if line:
                md_lines.append(f"- `{line}`")
    else:
        md_lines.append("- none")

    anomalies_md_path.write_text("\n".join(md_lines) + "\n", encoding="utf-8")

    return str(anomalies_json_path)


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Heuristic triage of a Freud run directory.",
        usage="freud/scripts/triage-run.sh <run_dir> [--top N]",
    )
    parser.add_argument("run_dir", help="Path to the Freud run directory")
    parser.add_argument("--top", type=int, default=20, help="Max top signals to keep")
    args = parser.parse_args()

    result = triage(args.run_dir, top_n=args.top)
    print(result)


if __name__ == "__main__":
    main()

File: freud/internal/analysis/triage.go
package analysis

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

var triagePatterns = []struct {
	ID    string
	Regex string
}{
	{"planner_output_repaired", "planner_output_repaired|output_repaired"},
	{"parse_failures", "non-parseable|failed to parse|parse fallback|parse error|parse failure"},
	{"forced_terminal", "forced terminal|forced_terminal_answer"},
	{"queue_saturation", "queue full|queue_saturation|step limit reached"},
	{"policy_denials", "action denied|denied by superego|policy denied"},
	{"provider_failures", "timeout|unavailable|provider check failed"},
	{"cache_divergence", "llm_cache_divergence"},
}

var ignoredSignalRegexes = []*regexp.Regexp{
	regexp.MustCompile(`(?i)scenario_id=\S+\s+selector=\S+$`),
	regexp.MustCompile(`(?i)status=pass\b`),
	regexp.MustCompile(`(?i)scenarios_total=.*scenarios_failed=0$`),
	regexp.MustCompile(`(?i)BUILD SUCCESSFUL`),
	regexp.MustCompile(`(?i)actionable tasks:`),
}

type patternResult struct {
	ID     string
	Count  int
	Sample string
}

// Triage runs heuristic triage on a run directory.
// Returns the path to the generated anomalies.json.
func Triage(runDir string, topN int) (string, error) {
	if !DirExists(runDir) {
		return "", fmt.Errorf("run directory does not exist: %s", runDir)
	}

	logsDir := filepath.Join(runDir, "logs")
	artifactDir := filepath.Join(runDir, "artifacts")
	if err := os.MkdirAll(artifactDir, 0o755); err != nil {
		return "", err
	}
	summaryJSON := filepath.Join(artifactDir, "summary.json")

	// Record patterns
	var results []patternResult
	for _, p := range triagePatterns {
		r := recordPattern(p.ID, p.Regex, logsDir)
		results = append(results, r)
	}

	// Write pattern-counts.tsv
	var tsvLines []string
	for _, r := range results {
		tsvLines = append(tsvLines, fmt.Sprintf("%s\t%d\t%s", r.ID, r.Count, r.Sample))
	}
	os.WriteFile(filepath.Join(artifactDir, "pattern-counts.tsv"),
		[]byte(strings.Join(tsvLines, "\n")+"\n"), 0o644)

	// Top signals & pressure
	var topLines, pressureLines []string
	firstFailLine := ""

	if DirExists(logsDir) {
		rawTop := RgSearch("error|exception|failed|warning", logsDir, true, true, ":> Task :")
		topLines = filterIgnoredSignals(NonEmptyLines(rawTop))
		if len(topLines) > topN {
			topLines = topLines[:topN]
		}

		rawPressure := RgSearch(`decision_pressure=[0-9]+\.[0-9]+`, logsDir, false, true, "")
		pressureLines = NonEmptyLines(rawPressure)

		rawFail := RgSearch("error|exception|failed|traceback|assert", logsDir, true, true, ":> Task :")
		failLines := filterIgnoredSignals(NonEmptyLines(rawFail))
		if len(failLines) > 0 {
			firstFailLine = failLines[0]
		}
	}

	os.WriteFile(filepath.Join(artifactDir, "top-signals.tsv"),
		[]byte(strings.Join(topLines, "\n")+"\n"), 0o644)
	os.WriteFile(filepath.Join(artifactDir, "pressure-signals.tsv"),
		[]byte(strings.Join(pressureLines, "\n")+"\n"), 0o644)

	// First failing trace snippet
	firstFailureText := ""
	if firstFailLine != "" {
		parts := strings.SplitN(firstFailLine, ":", 3)
		if len(parts) >= 2 {
			fileRef := parts[0]
			lineStr := parts[1]
			if FileExists(fileRef) {
				if lineNum, err := strconv.Atoi(lineStr); err == nil {
					ctx := contextAround(fileRef, lineNum, 2)
					firstFailureText = firstFailLine + "\n---\n" + ctx
				} else {
					firstFailureText = firstFailLine
				}
			} else {
				firstFailureText = firstFailLine
			}
		} else {
			firstFailureText = firstFailLine
		}
	}
	os.WriteFile(filepath.Join(artifactDir, "first-failing-trace.txt"),
		[]byte(firstFailureText), 0o644)

	// Pressure analysis
	maxPressureVal := ""
	maxPressureRef := ""
	firstPressure075 := ""
	firstPressure090 := ""

	pressureRe := regexp.MustCompile(`decision_pressure=([0-9]+\.[0-9]+)`)
	for _, line := range pressureLines {
		m := pressureRe.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		value := m[1]
		if maxPressureVal == "" || compareGT(value, maxPressureVal) {
			maxPressureVal = value
			maxPressureRef = line
		}
		if firstPressure075 == "" && compareGE(value, 0.75) {
			firstPressure075 = line
		}
		if firstPressure090 == "" && compareGE(value, 0.90) {
			firstPressure090 = line
		}
	}

	// Write anomalies.json
	generatedAt := UTCNowISO()
	anomaliesPath := filepath.Join(artifactDir, "anomalies.json")

	var patternEntries []string
	for _, r := range results {
		patternEntries = append(patternEntries, fmt.Sprintf(
			`    {"id":"%s","count":%d,"sample":"%s"}`,
			JSONEscape(r.ID), r.Count, JSONEscape(r.Sample)))
	}

	var topSignalEntries []string
	for _, line := range topLines {
		if line != "" {
			topSignalEntries = append(topSignalEntries, fmt.Sprintf(`    "%s"`, JSONEscape(line)))
		}
	}

	jsonLines := []string{
		"{",
		fmt.Sprintf(`  "generated_at": "%s",`, generatedAt),
		fmt.Sprintf(`  "run_dir": "%s",`, JSONEscape(runDir)),
		`  "pattern_counts": [`,
		strings.Join(patternEntries, ",\n"),
		"  ],",
		`  "pressure": {`,
		fmt.Sprintf(`    "max": {"value": "%s", "ref": "%s"},`, JSONEscape(maxPressureVal), JSONEscape(maxPressureRef)),
		fmt.Sprintf(`    "first_ge_075": "%s",`, JSONEscape(firstPressure075)),
		fmt.Sprintf(`    "first_ge_090": "%s"`, JSONEscape(firstPressure090)),
		"  },",
		fmt.Sprintf(`  "first_failing_trace": "%s",`, JSONEscape(firstFailureText)),
		`  "top_signals": [`,
		strings.Join(topSignalEntries, ",\n"),
		"  ]",
		"}",
	}
	os.WriteFile(anomaliesPath, []byte(strings.Join(jsonLines, "\n")+"\n"), 0o644)

	// Write anomalies.md
	runStatus := ExtractJSONString(summaryJSON, "status")
	md := []string{
		"# Anomaly Triage",
		"",
		fmt.Sprintf("- Run dir: `%s`", runDir),
		fmt.Sprintf("- Generated at: `%s`", generatedAt),
		"",
		"## Signal Counts",
	}

	if runStatus == "pass" {
		md = append(md, "",
			"Interpretation for pass runs:",
			"- These counts are informational signal hits, not failures by themselves.",
			"- On deterministic scenario packs, some hits can reflect exercised behaviors described in passing scenario output.",
			"")
	} else if runStatus == "fail" {
		md = append(md, "",
			"Interpretation for fail runs:",
			"- These counts are heuristic signal hits to guide triage, not a final diagnosis on their own.",
			"- Prioritize the first failed step and failing log references above these aggregate counts.",
			"")
	}

	for _, r := range results {
		md = append(md, fmt.Sprintf("- `%s`: %d", r.ID, r.Count))
		if r.Sample != "" {
			md = append(md, fmt.Sprintf("  sample: `%s`", r.Sample))
		}
	}

	md = append(md, "", "## Pressure Spikes")
	if maxPressureVal != "" {
		md = append(md, fmt.Sprintf("- max decision pressure: `%s`", maxPressureVal))
		md = append(md, fmt.Sprintf("  ref: `%s`", maxPressureRef))
	} else {
		md = append(md, "- max decision pressure: `none`")
	}
	if firstPressure075 != "" {
		md = append(md, fmt.Sprintf("- first >= 0.75: `%s`", firstPressure075))
	} else {
		md = append(md, "- first >= 0.75: none")
	}
	if firstPressure090 != "" {
		md = append(md, fmt.Sprintf("- first >= 0.90: `%s`", firstPressure090))
	} else {
		md = append(md, "- first >= 0.90: none")
	}

	md = append(md, "", "## First Failing Trace Snippet")
	if firstFailureText != "" {
		md = append(md, "```text", firstFailureText, "```")
	} else {
		md = append(md, "- none")
	}

	md = append(md, "", "## Top Signals")
	if len(topLines) > 0 {
		for _, line := range topLines {
			if line != "" {
				md = append(md, fmt.Sprintf("- `%s`", line))
			}
		}
	} else {
		md = append(md, "- none")
	}

	os.WriteFile(filepath.Join(artifactDir, "anomalies.md"),
		[]byte(strings.Join(md, "\n")+"\n"), 0o644)

	return anomaliesPath, nil
}

func recordPattern(id, regex, logsDir string) patternResult {
	if !DirExists(logsDir) {
		return patternResult{ID: id}
	}
	raw := RgSearch(regex, logsDir, true, true, "")
	lines := NonEmptyLines(raw)
	sample := ""
	if len(lines) > 0 {
		sample = lines[0]
	}
	return patternResult{ID: id, Count: len(lines), Sample: sample}
}

func filterIgnoredSignals(lines []string) []string {
	var result []string
	for _, line := range lines {
		ignored := false
		for _, re := range ignoredSignalRegexes {
			if re.MatchString(line) {
				ignored = true
				break
			}
		}
		if !ignored {
			result = append(result, line)
		}
	}
	return result
}

func contextAround(filepath string, lineNum, context int) string {
	data, err := os.ReadFile(filepath)
	if err != nil {
		return ""
	}
	allLines := strings.Split(string(data), "\n")
	start := lineNum - 1 - context
	if start < 0 {
		start = 0
	}
	end := lineNum + context
	if end > len(allLines) {
		end = len(allLines)
	}
	var numbered []string
	for i := start; i < end; i++ {
		numbered = append(numbered, fmt.Sprintf("  %d\t%s", i+1, allLines[i]))
	}
	return strings.Join(numbered, "\n")
}

func compareGT(a, b string) bool {
	af, err1 := strconv.ParseFloat(a, 64)
	bf, err2 := strconv.ParseFloat(b, 64)
	if err1 != nil || err2 != nil {
		return false
	}
	return af > bf
}

func compareGE(a string, threshold float64) bool {
	af, err := strconv.ParseFloat(a, 64)
	if err != nil {
		return false
	}
	return af >= threshold
}
